### DISCUSSION TITLE
[Announcement] LFRic Apps vn2.0 Release

### DISCUSSION THREAD

**Discussion by james-bruten-mo** [MetOffice/simulation-systems#290] [Announcements]
Dear All,

I'm pleased to announce the release of LFRic Apps and Core vn2.0. Many thanks to all developers and reviewers for getting this release together, particularly to those involved in releasing Psyclone 3 and setting up new software stacks. There are a number of technical changes in this release, some of which will require changes to working practices:

**Update to Psyclone 3.0**

LFRic is now using Psyclone 3.0. This includes some breaking changes. Please see [this post](https://github.com/MetOffice/simulation-systems/discussions/286) for details.

**Upgrade Macros**

LFRic is now setup to use Rose Upgrade Macros. These now need to be used if rose metadata in LFRic is updated. There are new [Working Practices](https://metoffice.github.io/simulation-systems/WorkingPractices/macros.html#upgrade-macros-in-lfric) now available. We will also organise some drop in sessions focused on macros in the new year. Please pay particular attention to existing changes not yet committed as these may require editing to include an upgrade macro.

**New Met Office Platforms**

Both Apps and Core are now setup to run on Azure Spice and the EX Quads. All Met Office developers should have access to these in this pre-UAT early access phrase and are encouraged to try them out as soon as possible. We will be keeping these test and kgo on them up to date, so changes will need to work on new platforms as well the old Spice and XC40s.

For both apps and core, the rose-stem test suites have been setup to run "Developer" and "All" groups on Azure Spice and EX1A if launched from Azure Spice, and on Old Spice and XC40s if launched from Old Spice. 

**LFRic Core Cylc 8 Only**

Post vn2.0 only the Cylc 8 suite is valid for LFRic Core. The cylc 7 suite will soon be deleted, but launching via "make test-suite" will produce an error message. Instead use the ` rose stem...` syntax.

Many thanks,
James, on behalf of Sim IT :christmas_tree:

**Comment by MichaelWhitall** [comment]
Hi, awesome thanks for the excellent work getting upgrade macros working LFRic :)

Just a couple of minor "gotchas" that I've hit, which others might want to know about when trying to write and use upgrade macros in lfric_apps:

a) The new script to update all the lfric_apps rose stem apps ($UMDIR/SimSys_Scripts/lfric_macros/apply_macros.py) doesn't work if like me you forget to export CYLC_VERSION=8 before running it.  It gives a cryptic error about not being able to find "metomi.rose.upgrade" (which the versions.py file tries to import stuff from).  But this problem goes away with CYLC_VERSION=8.

b) When you run the apply_macros.py script, _it overwrites the versions.py file you have just written, moving its contents into the individual application versions.py files, and deletes the macro you have written from the original file!_  So if like me you wanted to try running the apply_macros.py script to test if your upgrade macro works before actually committing the changes you made to versions.py, beware! (but if this happens, you can get your macro back again by copying it back from one of the various macros in applications).

c) If adding an "enum" type namelist setting using an upgrade macro (the ones that manifest as integer multi-option switches in the UM physics, but appear in the lfric_apps namelists as character strings), you have to put ' ' inside the " " in your macro commands to get it to appear correctly formatted in the namelist.  So e.g. an example from my ticket [lfric_apps:#467](https://code.metoffice.gov.uk/trac/lfric_apps/ticket/467), I originally put this in my upgrade macro:
`        self.add_setting(config, ["namelist:microphysics", "i_update_precfrac"], "homog")`
But this adds:
`i_update_precfrac=homog`
to the namelists, which doesn't work.  What we need to do is:
`        self.add_setting(config, ["namelist:microphysics", "i_update_precfrac"], "'homog'")`
which adds:
`i_update_precfrac='homog'`
which is now formatted correctly :)

Cheers!
Mike

**Comment by james-bruten-mo** [comment]
Hi Mike, many thanks for your comments and giving the macros a try - they're all reasonable points and I'll add some tips in the Working Practices. 

On the last point, I've dug a bit into UM/Jules macros to check I've not modified the behaviour. Needing to add the ' inside the " is how it worked before. There aren't many examples of that format of metadata, but one in Jules at least: https://code.metoffice.gov.uk/trac/jules/browser/main/trunk/rose-meta/jules-standalone/version63_70.py#L228 which is doing similar.

**Reply by MichaelWhitall** [reply]
That's great thanks James :)

Where I first noticed that my macro wasn't working (because of the missing ' inside ") was actually where the value of a new setting needed to be set depending on the value of an existing enum "cv_scheme" in the namelist:
[versions.py](https://code.metoffice.gov.uk/trac/lfric_apps/changeset/5785/main/branches/dev/michaelwhitall/vn2.0_improve_precfrac/science/um_physics_interface/rose-meta/um-microphysics/versions.py?old=5732&old_path=main%2Ftrunk%2Fscience%2Fum_physics_interface%2Frose-meta%2Fum-microphysics%2Fversions.py):
```
        # Set l_mcr_precfrac (switch for prognostic precip fraction);
        # So-far this has been hardwired in the code to be true if using
        # the comorph convection scheme, and false if not.
        # Replicate this when upgrading existing workflows from vn2.0,
        # to ensure they preserve answers
        nml = "namelist:convection"
        cv_scheme = self.get_setting_value(config, [nml, "cv_scheme"])
        if cv_scheme == "'comorph'":
            l_mcr_precfrac = ".true."
        else:
            l_mcr_precfrac = ".false."
        # end if

        # Add new microphysics settings to namelist
        nml = "namelist:microphysics"
        self.add_setting(config, [nml, "l_mcr_precfrac"], l_mcr_precfrac)
```

The line to test if the convection scheme used is comorph (`if cv_scheme == "'comorph'":`) doesn't work without the inner ' around 'comorph' (it did't raise an error or anything; it just silently evaluated to False when it was supposed to be True!)
